### PR TITLE
Flying Coaster track additions | Part 2

### PR DESCRIPTION
**Pull request by karst**
image

----------------------

Continuation of #16164 

The following PR will add new pieces to the B&M Flying Coaster.

* Large Inverted Half Loops
* Inverted Boosters

image

As a bonus this PR fixes a regression I found in my previous PR which also has to do with the flying coaster.

### CODE REVIEW

**Comment by SpartanFrederic104** [comment]
Oh you added more track color presets? That's actually amazing. It would be nice if other ride types like the Inverted Impulse Coaster and Vertical Drop Roller Coaster got more track color presets.
Anyway, the large inverted half loops still have the supports terminating a few feet before touching the actual track. How will you fix that so the supports in the large inverted half loops actually touch the track? Or is that just not possible?

**Comment by karst** [comment]
On the bottom of the PR i stated i still have to do the large loop supports haha. I will get to it, it's why I currently have it as a draft.

**Comment by karst** [comment]
I am wondering if it'd be nice to add strings for the new loops that would say inside or outside? 

If not this PR is ready to be reviewed.

**Comment by SpartanFrederic104** [comment]
If you want, then add the new text strings for those new inverted/uninverted loops. If not, that's fine.

**Comment by karst** [comment]
I think the strings might get too long It'd be something like `Large Half Looping (Right) (Outside)`

File: src/openrct2/ride/coaster/meta/LayDownRollerCoaster.h [hunk 1]
```diff
@@ -60,7 +60,7 @@ constexpr const RideTypeDescriptor LayDownRollerCoasterAltRTD =
     SET_FIELD(AlternateType, RIDE_TYPE_NULL),
     SET_FIELD(Category, RIDE_CATEGORY_NONE),
     SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_FLAT_ROLL_BANKING, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_HELIX_LARGE, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_BLOCK_BRAKES, TRACK_INLINE_TWIST_INVERTED, TRACK_FLYING_HALF_LOOP_INVERTED_DOWN, TRACK_CORKSCREW_INVERTED}),
-    SET_FIELD(ExtraTrackPieces, {TRACK_FLYING_HALF_LOOP_INVERTED_UP}),
+    SET_FIELD(ExtraTrackPieces, {TRACK_FLYING_HALF_LOOP_INVERTED_UP, TRACK_STATION_END}),
```
Comment (SpartanFrederic104) [inline src/openrct2/ride/coaster/meta/LayDownRollerCoaster.h:63]: Will you be adding the large steep slope pieces from the Corkscrew/Hypercoaster to the Lay-down too?

Comment (spacek531) [reply src/openrct2/ride/coaster/meta/LayDownRollerCoaster.h:63]: I forgot those existed. Probably better not to include them in this PR.

Comment (karst) [reply src/openrct2/ride/coaster/meta/LayDownRollerCoaster.h:63]: This is really easy to add with a single line. It's the same line I used to make the Looping Coaster draw the missing elements from the LIM Launch coaster. 
image

Comment (SpartanFrederic104) [reply src/openrct2/ride/coaster/meta/LayDownRollerCoaster.h:63]: The booster track piece is the last remaining track piece from the Corkscrew/Hypercoaster that isn't present on the Lay-down RC currently.

File: src/openrct2/ride/coaster/meta/FlyingRollerCoaster.h [hunk 2]
```diff
@@ -51,6 +51,9 @@ constexpr const RideTypeDescriptor FlyingRollerCoasterRTD =
         { COLOUR_OLIVE_GREEN, COLOUR_OLIVE_GREEN, COLOUR_TEAL },
         { COLOUR_TEAL, COLOUR_TEAL, COLOUR_BORDEAUX_RED },
         { COLOUR_BRIGHT_RED, COLOUR_BRIGHT_RED, COLOUR_DARK_PURPLE },
+        { COLOUR_SATURATED_RED, COLOUR_SATURATED_RED, COLOUR_DARK_YELLOW },
```
Comment (duncanspumpkin) [inline src/openrct2/ride/coaster/meta/FlyingRollerCoaster.h:54]: can you keep this PR to just the new pieces. Make this a separate PR

**Comment by spacek531** [comment]
Rebase fail?

**Review by Gymnasiast — CHANGES_REQUESTED** [review]
You probably rebased it against your outdated `develop` branch. Please rebase it against the upstream `develop`.

Also, the supports of the large half loops don’t reach the track properly. This needs to be fixed.

**Comment by karst** [comment]
Haha yea i rebased it and a day after another g2 update came across. I'll fix it tonight

**Comment by karst** [comment]
What still needs to happen with the large half loops? Is there something I am not seeing?
image
image

**Comment by SpartanFrederic104** [comment]
Fred's Workbench 2022-09-05 20-18-39
I believe there is one last track piece the flying coaster is missing from the B&M track: the small banked curve to gentle slope upward transition piece which is present on the Twister, Hyper-Twister, and Vertical Drop Coasters.

**Comment by karst** [comment]
Your wish is my command 🙏
image

File: src/openrct2/ride/coaster/LayDownRollerCoaster.cpp [hunk 3]
```diff
@@ -16492,5 +16492,5 @@ TRACK_PAINT_FUNCTION get_track_paint_function_lay_down_rc(int32_t trackType)
         case TrackElemType::FlyerHalfLoopUninvertedDown:
             return lay_down_rc_track_half_loop_uninverted_down;
     }
-    return nullptr;
+    return get_track_paint_function_corkscrew_rc(trackType);
```
Comment (Gymnasiast) [inline src/openrct2/ride/coaster/LayDownRollerCoaster.cpp:16495]: This is not related to the rest, could you revert it?

Comment (karst) [reply src/openrct2/ride/coaster/LayDownRollerCoaster.cpp:16495]: Do you wish me to seperate the lay-down changes to a seperate pr for ya?

File: src/openrct2/ride/coaster/meta/FlyingRollerCoaster.h [hunk 4]
```diff
@@ -65,8 +65,8 @@ constexpr const RideTypeDescriptor FlyingRollerCoasterAltRTD =
...
     SET_FIELD(AlternateType, RIDE_TYPE_NULL),
     SET_FIELD(Category, RIDE_CATEGORY_NONE),
-    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_LIFT_HILL, TRACK_FLAT_ROLL_BANKING, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_HELIX_LARGE, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_BLOCK_BRAKES, TRACK_SLOPE_ROLL_BANKING, TRACK_INLINE_TWIST_INVERTED, TRACK_FLYING_HALF_LOOP_INVERTED_DOWN, TRACK_QUARTER_LOOP_INVERTED_DOWN, TRACK_FLYING_LARGE_HALF_LOOP_INVERTED_DOWN}),
-    SET_FIELD(ExtraTrackPieces, {TRACK_FLYING_HALF_LOOP_INVERTED_UP}),
+    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_LIFT_HILL, TRACK_FLAT_ROLL_BANKING, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_HELIX_LARGE, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_BLOCK_BRAKES, TRACK_SLOPE_ROLL_BANKING, TRACK_INLINE_TWIST_INVERTED, TRACK_FLYING_HALF_LOOP_INVERTED_DOWN, TRACK_QUARTER_LOOP_INVERTED_DOWN, TRACK_FLYING_LARGE_HALF_LOOP_INVERTED_DOWN, TRACK_BOOSTER}),
```
Comment (Gymnasiast) [inline src/openrct2/ride/coaster/meta/FlyingRollerCoaster.h:68]: Do I understand correctly that you enabled this because it’s present on a real-life B&M Flyer?

File: src/openrct2/ride/coaster/FlyingRollerCoaster.cpp [hunk 5]
```diff
@@ -6554,8 +6554,31 @@ static void flying_rc_track_booster(
     }
     else
     {
-        // Should not occur, except when converting from other coaster types.
-        flying_rc_track_brakes(session, ride, trackSequence, direction, height, trackElement);
+        switch (direction)
+        {
+            case 0:
+            case 2:
+                PaintAddImageAsParentRotated(
```
Comment (Gymnasiast) [inline src/openrct2/ride/coaster/FlyingRollerCoaster.cpp:6561]: You can remove the switch:

```
const auto imageId = (direction & 1) ? SPR_G2_BM_INVERT_BOOSTER_2 : SPR_G2_BM_INVERT_BOOSTER_1;
                PaintAddImageAsParentRotated(
                    session, direction, session.TrackColours[SCHEME_TRACK] | imageId, { 0, 0, height + 24 },
                    { 32, 20, 1 }, { 0, 6, height + 22 });
```

File: src/openrct2/ride/coaster/FlyingRollerCoaster.cpp [hunk 6]
```diff
@@ -14594,22 +14617,22 @@ static void flying_rc_track_left_flyer_twist_up(
                 case 0:
                     PaintAddImageAsParentRotated(
                         session, direction, session.TrackColours[SCHEME_TRACK] | 27430, { 0, 6, height - 5 }, { 32, 20, 3 },
-                        { 0, 6, height + 8 });
+                        { 0, 6, height + 6 });
```
Comment (Gymnasiast) [inline src/openrct2/ride/coaster/FlyingRollerCoaster.cpp:14620]: This modifies existing code, which needs very good justification, could you elaborate?

File: src/openrct2/ride/coaster/meta/FlyingRollerCoaster.h (hunk 4 above)
Comment (karst) [reply src/openrct2/ride/coaster/meta/FlyingRollerCoaster.h:68]: I have enabled the Booster on the flyer since it's present on FLY. This is currently the only flying coaster that has an LSM launch. The new generation Vekoma flyers look a lot closer to the B&M flyer than they do to the lay-down coaster.

I have enabled this so people can build FLY like rollercoasters, which I think will be a fun gameplay addition. 

I have also made a texture for the LSM launch for it.

File: src/openrct2/ride/coaster/FlyingRollerCoaster.cpp (hunk 6 above)
Comment (karst) [reply src/openrct2/ride/coaster/FlyingRollerCoaster.cpp:14620]: I have made a mistake in a previous PR which makes the vehicle clip through the rails. This change will make sure that the bug that was fixed originally doesn't come back, but also that the vehicle no longer clips.

File: src/openrct2/ride/coaster/LayDownRollerCoaster.cpp (hunk 3 above)
Comment (Gymnasiast) [reply src/openrct2/ride/coaster/LayDownRollerCoaster.cpp:16495]: Yup.

File: src/openrct2/ride/coaster/LayDownRollerCoaster.cpp [hunk 7]
```diff
@@ -16492,5 +16492,5 @@ TRACK_PAINT_FUNCTION get_track_paint_function_lay_down_rc(int32_t trackType)
         case TrackElemType::FlyerHalfLoopUninvertedDown:
             return lay_down_rc_track_half_loop_uninverted_down;
     }
-    return get_track_paint_function_corkscrew_rc(trackType);
+    return nullptr;
```
Comment (Gymnasiast) [inline src/openrct2/ride/coaster/LayDownRollerCoaster.cpp:16495]: revert

File: src/openrct2/ride/coaster/meta/LayDownRollerCoaster.h [hunk 8]
```diff
@@ -19,8 +19,8 @@ constexpr const RideTypeDescriptor LayDownRollerCoasterRTD =
...
     SET_FIELD(Category, RIDE_CATEGORY_ROLLERCOASTER),
-    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_LIFT_HILL, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_BLOCK_BRAKES, TRACK_INLINE_TWIST_UNINVERTED, TRACK_FLYING_HALF_LOOP_UNINVERTED_UP, TRACK_CORKSCREW_UNINVERTED, TRACK_SLOPE_STEEP_LONG}),
-    SET_FIELD(ExtraTrackPieces, {TRACK_FLYING_HALF_LOOP_UNINVERTED_DOWN, TRACK_BOOSTER}),
+    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_LIFT_HILL, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_BLOCK_BRAKES, TRACK_INLINE_TWIST_UNINVERTED, TRACK_FLYING_HALF_LOOP_UNINVERTED_UP, TRACK_CORKSCREW_UNINVERTED}),
+    SET_FIELD(ExtraTrackPieces, {TRACK_FLYING_HALF_LOOP_UNINVERTED_DOWN}),
```
Comment (Gymnasiast) [inline src/openrct2/ride/coaster/meta/LayDownRollerCoaster.h:23]: revert

**Review by Gymnasiast — CHANGES_REQUESTED** [review]
There is a little glitch when trains go over the top of the loops (not visible on the screenshot, but this is the place you should look):

afbeelding

**Comment by karst** [comment]
If you are talking about the glitch where the cars are almost like bouncing below and above the track, this already happened before this PR (if you would let the trains run without a texture). I believe it only occurs on the top part of the large outside loop to the left. I have no clue how the vehicle sub positions work so @spacek531 would have to take a look a that in either a separate PR or I'll have to cherry pick it into here.

**Comment by Gymnasiast** [comment]
I’d like that to be fixed before this PR goes in (either here or in a PR of its own) - this PR is going to make that visible, after all.

Btw, it also happens on the top of the half loop that was added in @spacek531 ’s PR. I overlooked it, apparently.

afbeelding

**Comment by karst** [comment]
What I could do is make these track pieces still unavailable to be placed down (only renderable), then someone else can fix that issue before adding that back?

**Comment by Gymnasiast** [comment]
I’m reluctant to do so, as it makes it a bit too easy not to fix the actual issue.

**Comment by karst** [comment]
These refactors will be the bane of me 😂

**Comment by spacek531** [comment]
Issues related to glitching when transitioning between inverted and uninverted track elements are caused by the uncap FPS option. Seems like out of the scope of this PR.

**Comment by karst** [comment]
I have made a bug report #18309 for it. I think the main reason why this came up now is because this element is one of the most obvious ones for it to occur on.

File: distribution/changelog.txt [hunk 9]
```diff
@@ -1,6 +1,6 @@
 0.4.3 (in development)
 ------------------------------------------------------------------------
-- Feature: [##17782] The Flying Coaster now has access to boosters and can draw outside loops.
+- Feature: [#17782] The Flying Coaster now has access to boosters and can draw outside loops.
```
Comment (Gymnasiast) [inline distribution/changelog.txt:3]: \# is not a hashtag, it’s a hash. A hashtag is a hash followed by a tag, \#likethis.

</zeikmodus>

File: src/openrct2/sprites.h [hunk 10]
```diff
@@ -1293,6 +1293,9 @@ enum
     SPR_RIDE_DESIGN_PREVIEW_CLASSIC_WOODEN_ROLLER_COASTER_TRACK,
     SPR_RIDE_DESIGN_PREVIEW_CLASSIC_WOODEN_ROLLER_COASTER_SUPPORTS,
 
+    SPR_G2_BM_INVERT_BOOSTER_1,
+    SPR_G2_BM_INVERT_BOOSTER_2,
+
     SPR_G2_LIM_LAUNCHED_TRACK_BEGIN = SPR_RIDE_DESIGN_PREVIEW_CLASSIC_WOODEN_ROLLER_COASTER_SUPPORTS + 1,
```
Comment (Gymnasiast) [inline src/openrct2/sprites.h:1299]: This breaks the LIM launched coaster’s sprites. Thankfully, the g2 warning caught this. I’ll push a fix.

**Comment by Gymnasiast** [comment]
It also seems like the boosters are partly recoloured in the track’s second colour, rather than being white. Is that intentional?

afbeelding

**Comment by karst** [comment]
> It also seems like the boosters are partly recoloured in the track’s second colour, rather than being white. Is that intentional?
> 
> afbeelding

This is indeed intentional, the LSMs of FLY (which I based the sprite of) are hidden on the inside of metal sleeves (basically).

**Comment by karst** [comment]
> Tested it. Alignment seems a bit off here:
> 
> afbeelding
> 
> Could you look into that?
> 
> It seem you could probably lower the whole thing by a pixel or two and still have it connect properly on top.

I will take a look at it, I believe I have tried it before but the entirety of the bounding boxes were messed up. If that's the case I will have to redo those, but I'll check.

File: distribution/changelog.txt (hunk 9 above)
Comment (ocalhoun6) [reply distribution/changelog.txt:3]: Originally called an _octothorpe_. Also 'pound sign' and 'number sign'. Even calling it a "hash" is still an invention of Twitter's marketing department.

Comment (karst) [reply distribution/changelog.txt:3]: In dutch we call it a hekje aka "little fence"

**Comment by karst** [comment]
image

The top will end up looking like this, which is probably a tad cleaner, but still messy. 
The nightmare lies with the bounding boxes that get completely messed up by the singular pixel shift. it'll connect perfectly to the bottom 
image
but this is what you get in return for it 
image

It's why I originally decided on having the pixel be shifted off this tiny bit.

I should be able to tackle it if I'd have a tool like described in #18330, but until that exists it will be guess work and trial and error and ill can break my neck over that for days on end.

**Comment by spacek531** [comment]
I was the one that made the decision to have it be a little janky at the bottom. It disguises the jank better than moving it one pixel down and having the jank be at the top.

**Comment by Gymnasiast** [comment]
I guess we have no choice then. Ah well.

**Review by Gymnasiast — APPROVED** [review]
